Meta: Switch to clang-format-15 as the standard formatter

linusg opened the pull request:
Haven't tested this locally as I use Fedora, which ships with LLVM 15 already, but in theory using the updated apt repository should _just work_.

This finally fixes the remaining west `const`s.

alimpfard commented on AK/Atomic.h:
```diff
@@ -29,109 +29,109 @@ static inline void full_memory_barrier() noexcept
 
 template<typename T>
-static inline T atomic_exchange(volatile T* var, T desired, MemoryOrder order = memory_order_seq_cst) noexcept
+static inline T atomic_exchange(T volatile* var, T desired, MemoryOrder order = memory_order_seq_cst) noexcept
```
Unexpected but welcome change :^)

alimpfard commented on AK/BinaryBufferWriter.h:
```diff
@@ -19,7 +19,8 @@ class BinaryBufferWriter {
     template<typename T>
-    requires(AK::Detail::IsTriviallyConstructible<T>) T& append_structure()
+        requires(AK::Detail::IsTriviallyConstructible<T>)
+    T& append_structure()
```
Can we configure this? "requires on the same line as the rest of the function decl"?

linusg replied:
Reformatted with

```
RequiresClausePosition: WithFollowing
IndentRequiresClause: false
```

Still does a bunch of changes tho.

alimpfard replied:
Seems like the best of the available options, so :shrug:

linusg commented:
Waiting for our Arch friends' distro maintainers to do the thing.

rymiel commented:
Note that since clang-format-15 has requires expression and requires clause support the following workarounds can be removed:

https://github.com/SerenityOS/serenity/blob/2af028132adca1ae9fbb81a655a9ce00420d8ad4/AK/StdLibExtras.h#L31-L34
https://github.com/SerenityOS/serenity/blob/8a5d2be617a82c93cfafa6c714dff67ab5ff0f7b/AK/NonnullRefPtr.h#L203-L206
https://github.com/SerenityOS/serenity/blob/998c1152efa116da5fabf89363d178c8fb729d77/Kernel/Library/NonnullLockRefPtr.h#L213-L216
https://github.com/SerenityOS/serenity/blob/b2164ad979ef3dd0761d833dcc4f26154bf40b29/Meta/Lagom/Tools/CodeGenerators/LibUnicode/GeneratorUtil.h#L184-L189

Additionally, https://github.com/llvm/llvm-project/issues/56602 landed in 15 so the following can be removed:

https://github.com/SerenityOS/serenity/blob/849499988e3db08dea73f2444d91936710b91c6f/AK/Platform.h#L71-L72

Additionally, this can also be removed:
https://github.com/SerenityOS/serenity/blob/660d2b53b1206e868d5470eee80b5e62d7e30da7/Userland/Libraries/LibRegex/RegexParser.cpp#L2543-L2545
(I think that was fixed a couple versions ago actually)

Regarding this callout:
https://github.com/SerenityOS/serenity/blob/7398fd107c8e68d19f758d7346c1ea13c31b4826/AK/Concepts.h#L54

clang-format now has proper support for requires-clauses, but I don't think it can be configured to use a style like shown in that file (because i've never seen such style before). And there are a bunch of bugs with requires-clauses specifically that I've fixed but will only land in clang-format-16, so that can wait

supercomputer7 commented:
I said this before and I will say this again (as an Arch user) - please don't wait for the Arch people to get around this as this can take literally months before they will package-and-ship llvm-15. This is a sad situation for how Arch handles llvm updates, but we should not care about how they package or when they package their stuff.
What I will kindly ask is to provide a way for me to easily compile llvm-15 with clang-format (so me and other users can easily do that) and use that for Serenity. I am sure it's not an hard thing to do - we already compile clang for whoever that wants it, so I am sure generating clang-format-15 can be done without a problem.

supercomputer7 commented:
I looked in the `Toolchain` directory I had locally since I compiled clang about a month ago and I see that there's a `clang-format` binary already there, waiting to be used. Now I am convinced we should never consider how or when Linux distros ship their llvm packages - if we simply can compile and use the latest, then everything is really OK.
At least this is my opinion on the matter.

rymiel commented:
Yep! `clang-format` is directly part of the `clang` module, not as a separate project under i.e. `clang-tools-extra` (which seems to be built anyway, so doesn't matter)
> Also, please note that there's already llvm with version 16, so maybe we should use that instead.

Also also note that 16 is currently the version of the main branch, and not a released version. It will initially release at about March of next year, if my time-keeping is correct.
If that isn't a problem, though, 16 already has a bunch of clang-format fixes in it

linusg commented:
@supercomputer7 alright, rebased & will merge once CI is green. :)

@rymiel thanks for pointing all of these out! I added another commit removing the now redundant clang-format off comments.

rymiel commented on Userland/Libraries/LibRegex/RegexParser.cpp:
```diff
@@ -2541,7 +2541,6 @@ Optional<ECMA262Parser::PropertyEscape> ECMA262Parser::read_unicode_property_esc
     consume(TokenType::LeftCurly, Error::InvalidPattern);
 
     // Note: clang-format is disabled here because it doesn't handle templated lambdas yet.
```
This comment too?

linusg replied:
yep :+1:

ADKaster commented on Userland/Applications/Spreadsheet/Readers/XSV.h:
```diff
@@ -107,7 +107,7 @@ class XSV {
         size_t size() const { return m_xsv.headers().size(); }
 
-        using ConstIterator = AK::SimpleIterator<const Row, const StringView>;
+        using ConstIterator = AK::SimpleIterator<const Row, StringView const>;
```
curiously, it didn't format this as `AK::SimpleIterator<Row const, StringView  const>`. nor below in the constructor of RowIterator, `RowIterator(const XSV&`

linusg replied:
Apparently there's still 100+ of these, sigh. Will not fix those up by hand, that's for the clang-format people to fix.

![image](https://user-images.githubusercontent.com/19366641/204536909-f45010fd-f1ad-4ad5-b89b-d140e6d7bf3c.png)

ADKaster commented on Userland/Libraries/LibCodeComprehension/Cpp/Tests/complete_includes.cpp:
```diff
@@ -1,5 +1,4 @@
 
-void foo() {}
-
+void foo() { }
```
hm I thought cpp tests would be omitted from clang-format, is that just the ones in Tests/LibCPP? also, why is this test here and not in Tests/ 🤔

linusg replied:
Good catch, added a commit similar to https://github.com/SerenityOS/serenity/commit/42865b8975818116b922aa2411eb167618fc532c and reverted this change

No idea about the second question :)

ADKaster commented on Userland/Libraries/LibEDID/EDID.h:
```diff
@@ -439,11 +439,11 @@ class Parser final {
     template<typename T>
     requires(IsIntegral<T> && sizeof(T) > 1) T read_le(T const*)
-    const;
+        const;
```
Does it really want the `const` here on its own line for these templated member functions? Also in .cpp file.

linusg replied:
That's what the blind reformat did, manually adding a line break after the requires clause and running the formatter again fixed it.

ADKaster commented on Userland/Libraries/LibJS/AST.h:
```diff
@@ -192,19 +192,19 @@ class ExpressionStatement final : public Statement {
-        }
-        -> SameAs<ThrowCompletionOr<void>>;
-};
+concept ThrowCompletionOrVoidFunction = requires(Func func, Args... args) {
+                                            {
+                                                func(args...)
+                                                }
+                                                -> SameAs<ThrowCompletionOr<void>>;
+                                        };
```
This is a disaster,  but oddly the clang-format-14 version was a disaster as well. So meh.

rymiel replied:
(will be fixed in clang-format-16, nothing to do for now)

ADKaster commented on Userland/Libraries/LibPDF/Encoding.h:
```diff
@@ -421,7 +421,7 @@
     V("ϑ", theta1, 74)                                           \
     V("3", three, 51)                                            \
-    V("", trademarksans, 228)                                   \
+    V("", trademarksans, 228)                                 \
```
Ooo now *these* are odd. Why did it decide to indent some of these macros with unicode characters less than the surrounding lines? counting bytes instead of grapheme clusters are we, clang-format?

rymiel replied:
Those are all Private Use Area characters. AFAIK clang *should* respect character width, but I assume PUA ones are just ignored

ADKaster commented on Userland/Libraries/LibTest/TestCase.h:
```diff
@@ -42,34 +42,43 @@ void add_test_case_to_suite(NonnullRefPtr<TestCase> const& test_case);
+    static void __BENCHMARK_FUNC(x)();                                                               \
+    struct __BENCHMARK_TYPE(x) {                                                                     \
+        __BENCHMARK_TYPE(x)                                                                          \
+        ()                                                                                           \
```
clang-format really doesn't know what to do with these constructors, does it :D macro overload. no change requested.

ADKaster reviewed (commented):
I noticed a couple things, no deal breakers though.

rymiel commented on Userland/Utilities/test-pthread.cpp:
```diff
@@ -31,8 +31,11 @@ static ErrorOr<void> test_once()
         threads.last().start();
     }
+    // clang-format off
+    // It wants to put [[maybe_unused]] on its own line, for some reason.
```
I've opened https://github.com/llvm/llvm-project/issues/59235 for this

rymiel replied:
Should have tested more.
Basically it seems that annotations following these control flow statements without braces are somewhat ambiguous in C++, and it only works here because of the declaration (`auto res`).
But because of the declaration, I don't even think maybe_unused is strictly required *by the compiler* (which doesn't mean it shouldn't be there at all to communicate something to the reader)
